### PR TITLE
Megatron hidden transformations

### PR DESCRIPTION
**Pull request by michalivne**
# What does this PR do ?

This PR adds support in applying transformation to the hiddens in Megatron Encoder-Decoder model, and augmenting a corresponding loss. This enables support in latent variable models, like VAE and MIM.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can add the following to your config :
hidden_transforms: [MegatronGaussianHiddenTransform()] and it will apply a gaussian transform to the hiddens.

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 1]
```diff
@@ -250,7 +250,7 @@ def loss_func(output_tensor):
                     lm_loss = loss_dict['lm loss']
                     loss = lm_loss
                     reduced_loss = average_losses_across_data_parallel_group([loss, lm_loss])
-                return loss, {'avg': reduced_loss}
+                return loss, {'loss': reduced_loss}
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:253]: Why are we changing 'avg' to 'loss' ?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 2]
```diff
@@ -919,7 +961,7 @@ def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int]
         logging.info(f"response: {response}")
         return response
 
-    def encode(self, tokens_enc, enc_mask, encoder_input=None, reconfigure_microbatch=True):
+    def encode(self, tokens_enc, enc_mask, encoder_input=None, batch_data=None, reconfigure_microbatch=True):
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:964]: Could you add `batch_data` to the doc string here?

File: nemo/collections/nlp/modules/common/megatron/transformations/megatron_hidden_loss.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        """
+        We expect input shapes to be [S x B x H] for Sequence, Batch, Hidden sizes (due to tensor parallel support).
+        We return a dictionary with dimensions [B x S x H], [B x S], [B], or [].
+
+        Implement your own loss calculations. Must return "loss" key.
+        loss shape - [B x S] for Batch, Sequence sizes
+        batch_data - a dictionary of additional data that can be used to calculate loss
+        """
+        raise NotImplementedError("Please implement loss calculations in child class")
+        # return {"loss": 0.0}
```
Comment (ericharper) [inline nemo/collections/nlp/modules/common/megatron/transformations/megatron_hidden_loss.py:62]: delete comment

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:253]: `avg` is not informative. It implicitly relates to `avg_loss` but since I can return multiple items with hidden transformations (all are averaged) I made it explicit. It does deviate from the GPT convention which returns `avg`. 
If you would like to have a unified convention I would recommend that I will also update the GPT models to return the same key.
Does that make sense?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py (hunk 2 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:964]: Done.

>

File: nemo/collections/nlp/modules/common/megatron/transformations/megatron_hidden_loss.py (hunk 3 above)
Comment (michalivne) [reply nemo/collections/nlp/modules/common/megatron/transformations/megatron_hidden_loss.py:62]: Done, also updated comments.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:253]: Kept as is, as discussed.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
